libdnf/hy-iutil.cpp

```diff
@@ -182,6 +183,106 @@ int checksum_write(const unsigned char *cs, FILE *fp)
     return 0;
 }
 
+static std::array<char, solv_userdata_solv_toolversion_size>
+get_padded_solv_toolversion()
+{
+    std::array<char, solv_userdata_solv_toolversion_size> padded_solv_toolversion{};
+    std::string solv_ver_str{solv_toolversion};
+    std::copy(solv_ver_str.rbegin(), solv_ver_str.rend(), padded_solv_toolversion.rbegin());
+
+    return padded_solv_toolversion;
+}
+
+int
+solv_userdata_fill(SolvUserdata *solv_userdata, const unsigned char *checksum, GError** error)
+{
+    if (strlen(solv_toolversion) > solv_userdata_solv_toolversion_size) {
+        g_set_error(error, DNF_ERROR, DNF_ERROR_INTERNAL_ERROR,
+                    _("Libsolv's solv_toolversion is: %zu long but we expect max of: %zu"),
+                    strlen(solv_toolversion), solv_userdata_solv_toolversion_size);
+        return 1;
+    }
+
+    // copy dnf solv file magic
+    memcpy(solv_userdata->dnf_magic, &(solv_userdata_magic[0]), solv_userdata_magic.size());
+
+    // copy dnf solv file version
+    memcpy(solv_userdata->dnf_version, &(solv_userdata_dnf_version[0]), solv_userdata_dnf_version.size());
+
+    // copy libsolv solv file version
+    memcpy(solv_userdata->libsolv_version, &(get_padded_solv_toolversion()[0]), solv_userdata_solv_toolversion_size);
+
+    // copy checksum
+    memcpy(solv_userdata->checksum, checksum, CHKSUM_BYTES);
+
+    return 0;
+}
+
+
+std::unique_ptr<SolvUserdata>
+solv_userdata_read(FILE *fp)
+{
+    unsigned char *dnf_solvfile_userdata_read = NULL;
+    int dnf_solvfile_userdata_len_read;
+    if (!fp) {
+        return nullptr;
+    }
+
+    int ret_code = solv_read_userdata(fp, &dnf_solvfile_userdata_read, &dnf_solvfile_userdata_len_read);
+    // The userdata layout has to match our struct exactly so we can just cast the memory
+    // allocated by libsolv
+    std::unique_ptr<SolvUserdata> uniq_userdata(reinterpret_cast<SolvUserdata *>(dnf_solvfile_userdata_read));
+    if(ret_code) {
+        g_debug("Failed to read solv userdata: solv_read_userdata returned: %i", ret_code);
```

Comment:
This and similar errors (besides the normal case of checksum mismatch) should be logged higher than debug, because even though dnf can keep working, it's still an error state that probably needs addressing. I think these should be logged as errors in general, but I think `g_error()` actually calls `terminate()` IIRC? In that case I suppose log as warning...

Reply:
Yea `g_error()` does terminate but there is also `g_critical()` I have used that in a couple places. For the cases where the dnf/libsolv solvfile version doesn't match I used `g_warning()` since while rare it is not an error.